### PR TITLE
Allow multiplication in duration input value

### PR DESCRIPTION
**Pull request by lukaszachy**
To make it easier for adjust the multiplication happens last: No need to pay attention how adjust creates the value.

Only late I realized the input value for multiplication should accept float, so I will modify that later.
For now: does this approach will be accepted? It is preparation for #1843 so "adjust" doesn't need special handling of duration key.

Pull Request Checklist

* [x] implement the feature
* [x] write the documentation
* [x] extend the test coverage
* [x] modify the json schema
* [x] mention the version
* [x] include a release note

### CODE REVIEW

File: tmt/utils.py [hunk 1]
```diff
@@ -3323,19 +3323,29 @@ def shell_variables(
...
         'm': 60,
         'h': 60 * 60,
         'd': 60 * 60 * 24,
         }
-    if re.match(r'^(\d+ *?[smhd]? *)+$', str(duration)) is None:
+    # Couldn't create working validation regexp to accept '2 1m4'
+    # thus fixing the string so \b works well
+    fixed_duration = re.sub(r'([smhd])(\d)', r'\1 \2', str(duration))
+    re_validate = r'^((\*\s*\d+(?!\s*[smhd])\b\s*)|(\d+\s*[smhd]?\b)\s*)+$'
+    re_split = r'(?:(?:(\*)\s*(\d+)(?!\s*[smhd])\b\s*)|(?:(\d+)\s*([smhd]?)\b)\s*)'
```
Comment (happz) [inline tmt/utils.py:3337]: Would you consider verbose regular expressions? I can't tell whether the pattern is correct or not without diving deeply into dissecting it (tests would fail, so it probably is, but even the test might be wrong...). Using a verbose approach might improve readability, one can comment on what each component means, e.g. https://github.com/teemtee/tmt/blob/main/tmt/hardware.py#L116:

```python
CONSTRAINT_VALUE_PATTERN = re.compile(rf"""
    ^                                   # must match the whole string
    (?P<operator>{_OPERATOR_PATTERN})?  # optional operator
    \s*                                 # operator might be separated by white space
    (?P<value>.+?)                      # actual value of the constraint
    \s*                                 # there might be trailing white space
    $                                   # must match the whole string, I said :)
    """, re.VERBOSE)
```

Comment (lukaszachy) [reply tmt/utils.py]: changed in dd7a4621

**Review by happz — APPROVED** [review]

**Comment by happz** [comment]
@lukaszachy do you plan to mention the version and/or include a release note?

**Comment by lukaszachy** [comment]
> @lukaszachy do you plan to mention the version and/or include a release note?

I wasn't but it is a very good point.

**Comment by happz** [comment]
> > @lukaszachy do you plan to mention the version and/or include a release note?
> 
> I wasn't but it is a very good point.

1. those were the last two unchecked boxes on the PR checklist, and 2. this might be a good feature to point out in release notes, so I asked.

File: spec/tests/duration.fmf [hunk 2]
```diff
@@ -7,9 +7,12 @@ story:
 description:
     In order to prevent stuck tests consuming resources we define a
     maximum time for test execution. If the limit is exceeded the
-    running test is killed by the test harness. Use the same
-    format as the ``sleep`` command. Must be a ``string``. The
-    default value is ``5m``.
+    running test is killed by the test harness. Value extends the
+    format of the ``sleep`` command by allowing multiplication (``*[float]``).
+    First all time values are sum together and only after that are multiplied.
```
[the remark this answers is not in the document]
Comment (thrix) [reply spec/tests/duration.fmf:12]: Maybe even

```suggestion
First, all time values are summed together and then multiplied.
```

File: tests/unit/test_utils.py [hunk 3]
```diff
@@ -317,7 +317,7 @@ def create_workdir():
 
 
 def test_duration_to_seconds():
-    """ Check conversion from sleep time format to seconds """
+    """ Check conversion from  extended sleep time format to seconds """
```
Comment (thrix) [inline tests/unit/test_utils.py:320]: Double space

```suggestion
    """ Check conversion from extended sleep time format to seconds """
```

**Review by janhavlin — APPROVED** [review]

File: spec/tests/duration.fmf (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukaszachy) [reply spec/tests/duration.fmf:12]: https://quillbot.com/grammar-check hinted to use `First, all time values are summed together, and only after that are they multiplied.`

File: tmt/utils.py [hunk 4]
```diff
@@ -3337,19 +3338,52 @@ def shell_variables(
...
+                    (?P<suffix>[smhd])? # suffix
+                    \s*
+                )
+            )\b # Needs to end with word boundary to avoid splitting
+        '''
+    re_validate = re.compile(r'''
+        ^(  # Match beginning, opening of input group
+        ''' + raw_groups + r'''
+        \s*  # Optional spaces in the case of multiple inputs
+        )+$ # Inputs can repeat
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/utils.py:3373]: Isn't it better to remove one space from 3372 instead?

**Comment by lukaszachy** [comment]
ah, seems schema has the regexp as well, I need to change it.

File: spec/tests/duration.fmf (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukaszachy) [reply spec/tests/duration.fmf]: see 7445c842

File: tmt/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/utils.py:3382]: see https://github.com/teemtee/tmt/commit/7445c8423818203017db2396733e5d8e5b11d27e

**Review by thrix — APPROVED** [review]

**Comment by LecrisUT** [comment]
Sorry for commenting late on this closed PR, but I think `pint` would be a more robust approach:
```python
>>> import pint
>>> pint.Quantity("30 * 4 min")
<Quantity(120, 'minute')>
```

Some care needs to be made with regards to translating `30m` -> `30min`, but that can be done by configuring specific `pint.UnitRegistry.context` and overwriting the default values for `meter` and such.

**Comment by happz** [comment]
> Sorry for commenting late on this closed PR, but I think `pint` would be a more robust approach:
> 
> ```python
> >>> import pint
> >>> pint.Quantity("30 * 4 min")
> <Quantity(120, 'minute')>
> ```
> 
> Some care needs to be made with regards to translating `30m` -> `30min`, but that can be done by configuring specific `pint.UnitRegistry.context` and overwriting the default values for `meter` and such.

That's right, I didn't think about using Pint for this task. I'm not sure if it would handle the ordering of operations @lukaszachy implemented though, it does not follow the mathematical ordering of multiplication and addition.

**Comment by LecrisUT** [comment]
Indeed, but the current one is also a bit hard to follow. It can look good with
```yaml
    adjust:
        duration+: 1h
```
but the mathematical alternative is not bad either though:
```yaml
    adjust:
        duration+: + 1h
```

I was hoping that `pint` would fail if the syntax is broken, but alas it does not:
```python
>>> pint.Quantity("1h 40min")
<Quantity(40, 'hour * minute')>
>>> pint.Quantity("* 40min 1h")
Traceback (most recent call last):
```

At least the first issue is straightforward to verify by adding a check for `dimensionality`
```python
>>> pint.Quantity("1h 40min").dimensionality["[time]"]
2
```
